Add Accessibility Information and Best Practices to Doc Site

talbs opened the pull request:
**This work aims to tackle the following:**
* Adds in an informational page on Font Awesome (and icon fonts) and web accessibility - initially including background, use cases, and more information.
* Adds accessibility-minded examples to the examples page
* Updates the markup displayed on an icon detail page to use `aria-hidden="true"` with a link to more information on the new accessibility page
* Generally updates the Font Awesome site with some of these best practices.

**What this doesn't cover**
* adding or using [advanced icon fallback techniques](https://www.filamentgroup.com/lab/bulletproof_icon_fonts.html) for high contrast mode
* Updates to test-based files that are not end user facing

- - -

#### Reviewers

* [x] Copy and UI - @davegandy
* [ ] a11y - @dylanb

- - -

FYI, @tagliala. I'll comment on the few open GH issues that I'm hoping this work addresses as we progress. With the PR process in mind, would you recommend asking those members to review/take a look at this work?

tagliala commented:
@talbs thanks. The most important one should be #6133

dylanb commented on src/_includes/examples/accessible.html:
```diff
@@ -0,0 +1,96 @@
+    <div class="col-md-3 col-sm-4">
+      <p>
+        <a class="btn btn-default" href="path/to/settings">
+          <i class="fa fa-cog" aria-hidden="true"><span class="sr-only">Settings</span></i>
```
adding `sr-only` to something whose parent is has `aria-hidden="true"` will render it invisible to screen readers. You can fix this in two ways. Either (preferred)
```
<a class="btn btn-default" href="path/to/settings" title="Settings">
          <i class="fa fa-cog" role="presentation"></i>
</a>
```
or (lots of redundancy)
```
<a class="btn btn-default" href="path/to/settings" title="Settings">
          <i class="fa fa-cog" aria-hidden="true"><span class="sr-only">Settings</span></i>
</a>
```
This will also ensure that the anchor will work for systems like Dragon

dylanb commented on src/_layouts/icon.html:
```diff
@@ -49,10 +49,15 @@ <h1 class="info-class">
 {% endhighlight %}
-        <br>
-        <div class="lead"><i class="fa fa-info-sign"></i> Looking for more? Check out the <a href="{{ page.relative_path }}examples/">examples</a>.</div>
+        <p class="margin-bottom-lg">
```
role="presentation" is more semantically correct that aria-hidden="true"

talbs replied:
Should be resolved.

dylanb commented on src/_includes/brand-adblock-warning.html:
```diff
@@ -1,4 +1,4 @@
-<h4><i class="fa fa-warning"></i> Warning!</h4>
+<h4><i class="fa fa-warning" aria-hidden="true"></i> Warning!</h4>
```
see previous comment re: role="presentation"

talbs replied:
Should be resolved.

dylanb commented on src/_includes/accessibility/using-with-accessibility.html:
```diff
@@ -0,0 +1,78 @@
+
+  <h3>Icons used for pure decoration or visual styling</h3>
+  <p>
+    If you're using an icon to add some extra decoration or branding, it does not need to be announced to users as they are navigating your site or app aurally. Additionally, if you're using an icon to visually re-emphasize or add styling to content already present in your HTML, it does not need to be repeated to an assistive technology-using user. You can make sure this is not read by adding the ...
```
cf. previous comment re: role="presentation"

dylanb commented on src/_includes/accessibility/using-with-accessibility.html:
```diff
@@ -0,0 +1,78 @@
+
+  <h3>Icons that with semantic or interactive purpose</h3>
+  <p>
+    If you're using an icon to convey meaning (rather than only as a decorative element), ensure that this meaning is also conveyed to assistive technologies – for instance, include additional content that's visually hidden to users but semantically avaiable to screen readers. This goes for content you're abbreviating via icons as well as interactive controls (buttons, form elements, toggles, etc ...
```
it would be better to use the title attribute because of the following advantages

1. It works with all modern screen readers
2. It works well with the most ubiquitous speech recognition program (Dragon)
3. It help users with cognitive disabilities by providing a mouse tooltip

talbs commented:
FYI, @cptvitamin. Here's the PR I mentioned to bring Font Awesome's info and examples up to speed with good a11y practices. Mind taking a look?

cptvitamin commented on src/_includes/accessibility/background.html:
```diff
@@ -0,0 +1,15 @@
+
+  <ul>
+    <li>
+      The assistive technology may not finding any content to read out to a user
```
finding -> find

talbs commented:
@dylanb, can you walk me through why and exactly when would using a title attribute be better to recommend than the `<span class="sr-only">alternative text equivalent</span>`?

I want to make sure we have the right syntax for the right cases in our examples. Is there ever a case you'd want to use  `<span class="sr-only">alternative text equivalent</span>` rather than the title attribute suggestion you had?

talbs commented:
@dylanb, I think I addressed all of your feedback. Mind taking another look as well as chiming in on my questions above (https://github.com/FortAwesome/Font-Awesome/pull/8879#issuecomment-205370942)?

dylanb commented:
@talbs Definitely looks much better for screen readers

talbs commented:
> Apparently role="presentation" is wrong (#6133 (comment)) I would go only for aria-hidden="true" as before

Thanks, @tagliala. I'll revert back to using the `aria-hidden="true"` technique rather than the `role="presentation"` route.

talbs commented:
@dylanb, thanks for the continued help. Mind taking a hopefully final look and giving this work your :+1:  if you think its ready for primetime?

dylanb commented on src/_includes/community/requesting-new-icons.html:
```diff
@@ -23,7 +23,7 @@ <h2 class="page-header">Requesting New Icons</h2>
     </li>
     <li>
-      Request concrete objects: it's harder to make an icon to represent happiness, it's easier to make a smiley face.  <i class="fa fa-smile-o"></i>
+      Request concrete objects: it's harder to make an icon to represent happiness, it's easier to make a smiley face.  <i class="fa fa-smile-o" aria-hidden="true"></i>
```
I would argue that this example is not purely decorative and therefore needs an alt text

dylanb commented on src/_includes/examples/animated.html:
```diff
@@ -9,11 +9,11 @@ <h2 class="page-header">
-        <i class="fa fa-refresh fa-spin fa-3x fa-fw margin-bottom"></i>
-        <i class="fa fa-cog fa-spin fa-3x fa-fw margin-bottom"></i>
-        <i class="fa fa-spinner fa-pulse fa-3x fa-fw margin-bottom"></i>
+        <i class="fa fa-spinner fa-spin fa-3x fa-fw margin-bottom" aria-hidden="true"></i>
```
This list of icons is communicating that something is busy, they should have alt texts

talbs commented:
@dylanb, your suggestion to use `role="presentation"` for any icons that are purely decorative is really appealing, since its semantic and more straight-forward. **Can you confirm that `role="presentation"` is enough to hide all CSS-generated content (we use the `:before` and `:after` pseudo-elements to help render our icons) from assistive technology?**

If that's solid enough, we can rely on that in general rather than `aria-hidden="true"`.

dylanb commented:
I apologize for the confusion: role="presentation" is not going to hide content. When I made that suggestion, I had not yet internalized the desire to hide the unicode character and was interpreting this as a desire to hide any semantics on the `<i>` tag.

To hide content from screen readers, you must use `aria-hidden="true"`

dylanb commented:
ok, looking through your most recent changes, these have made things a mixture of better and worse and its probably my fault for not explaining adequately. I have created three bullet points below to summarize the salient information. If you want me to turn this into examples, please let me know.

1. `aria-hidden="true"` will hide all content and semantics and is the technique that should be used to hide unwanted unicode characters
1. `title` attribute should only be used as an accessible name on interactive elements like links or buttons, in all other cases, `sr-only` should be used in addition to title to support users with cognitive disabilities (see additional information below)
1. role="presentation" could be used on the `<i>` tags in cases where the unicode character is meaningful. This is really a "futureproofing" best practice because most screen reader settings do not cause the semantics of that tag to be exposed (but they could).

Additional Information: the title attribute is part of the ARIA name calculation algorithm as the "fallback". ARIA name calculation, while well supported for well known role combinations, is not supported by any browser+AT combinations in its entirety. So for example, using aria-label on a `<div>` will not work on most platforms, while using it on an interactive element like a button or a link, will work everywhere. The same applies to the title attribute. Therefore, title will provide the most bang for the buck when used on an interactive element, where it can serve as the accessible name, the visible hint and the target for Dragon. When used on non-interactive elements, it is really only serving the purpose of a hint which helps users with cognitive disabilities.

talbs commented:
>
1. `aria-hidden="true"` will hide all content and semantics and is the technique that should be used to hide unwanted unicode characters
1. `title` attribute should only be used as an accessible name on interactive elements like links or buttons, in all other cases, `sr-only` should be used in addition to `title` to support users with cognitive disabilities (see additional information below)
1. role="presentation" could be used on the `<i>` tags in cases where the unicode character is meaningful. This is really a "futureproofing" best practice because most screen reader settings do not cause the semantics of that tag to be exposed (but they could).

Perfect! Thanks, @dylanb. That's exactly the feedback I was looking for. I'll sync up our examples, sample icon code, and doc site markup to represent that. Expect 1 more round of spam shortly, :)

talbs commented:
A bit of status, all.

I've updated the examples, a11y page, and the Font Awesome site's markup to reflect the most recent use cases of icons and accessibility per @dylanb. Also, all commits have been squashed.

We've also decided to roll this work into a version bump of Font Awesome, which has a few more a11y-based goodies. With that said, once the dust settles here, I'll close this thread and open a PR to merge this into that version's branch.

@dylanb and @davegandy, is this good to go from your perspectives?

talbs commented:
Since this seems solid from both a11y and doc/UI points of view, I'm closing this PR as this work is being rolled into the [WIP 4.6.0 branch focusing in accessibility](https://github.com/FortAwesome/Font-Awesome/commits/4.6.0-wip).

Thanks for all of the help, folks. Really excited to get these improvements out to our users. :)

talbs commented:
Also, I'll leave this branch open until the `4.6.0` version goes live. **If you have any further feedback, please provide it here and I'll address it.**
